concat.js

```diff
@@ -80,27 +81,67 @@ Concat.inputNodesForConcatStats = function(inputNode, id, outputFile) {
   ];
 };
 
-Concat.prototype.shouldBuild = function() {
+Concat.prototype.calculatePatch = function() {
   var currentTree = this.getCurrentFSTree();
-  var isRebuild = !!this._lastTree;
-  var patch;
-
-  if (isRebuild) {
-    patch = this._lastTree.calculatePatch(currentTree);
-  }
+  var patch = this._lastTree.calculatePatch(currentTree);
 
   this._lastTree = currentTree;
 
-  // We build if this is an initial build (not a rebuild)
-  // or if the patch has non-zero length
-  return !isRebuild || patch.length !== 0;
+  return patch;
 };
 
 Concat.prototype.build = function() {
-  if (!this.shouldBuild()) {
+  var patch = this.calculatePatch();
+
+  // We skip building if this is a rebuild with a zero-length patch
+  if (patch.length === 0 && this._hasBuilt) {
     return;
   }
 
+  this._hasBuilt = true;
+
+  if (this.Strategy.isPatchBased) {
+    return this._doPatchBasedBuild(patch);
+  } else {
+    return this._doLegacyBuild();
+  }
+};
+
+Concat.prototype._doPatchBasedBuild = function(patch) {
+  if (!this.concat) {
+    this.concat = new this.Strategy(merge(this.sourceMapConfig, {
+      separator: this.separator,
+      inputDir: this.inputPaths[0],
+      header: this.header,
+      headerFiles: this.headerFiles,
+      footerFiles: this.footerFiles,
+      footer: this.footer,
+      allowNone: !!this.allowNone
+    }));
+  }
+
+  for (var i = 0; i < patch.length; i++) {
+    var operation = patch[i];
+    var method = operation[0];
+    var file = operation[1];
+
+    switch (method) {
+      case 'create':
+        this.concat.addFile(file);
+        break;
+      case 'unlink':
+        this.concat.removeFile(file);
+        break;
+      case 'change':
+        this.concat.updateFile(file);
+        break;
+    }
+  }
+
+  this.concat.write(path.join(this.outputPath, this.outputFile));
```

Comment:
I think inverting control is better. I feel the fewer concerns we put in the strategy, the easier it will be to swap maintain the different strategies.

Will update this shortly.